java/src/com/github/bazel_contrib/contrib_rules_jvm/junit5/JUnit5Runner.java

```diff
@@ -13,32 +13,68 @@ public class JUnit5Runner {
   private static final String JUNIT5_RUNNER_CLASS =
       "com.github.bazel_contrib.contrib_rules_jvm.junit5.ActualRunner";
 
+  private static final String JAVA17_SYSTEM_EXIT_TOGGLE =
+      "com.github.bazel_contrib.contrib_rules_jvm.junit5.Java17SystemExitToggle";
+
   public static void main(String[] args) {
     var testSuite = System.getProperty("bazel.test_suite");
 
+    var systemExitToggle = getSystemExitToggle();
+
     if (testSuite == null || testSuite.isBlank()) {
       System.err.println("No test suite specified");
-      exit(2); // Same error code as Bazel's own test runner
+      exit(systemExitToggle, 2); // Same error code as Bazel's own test runner
     }
 
     detectJUnit5Classes();
 
+    systemExitToggle.prevent();
+
     try {
       var constructor =
           Class.forName(JUNIT5_RUNNER_CLASS).asSubclass(RunsTest.class).getConstructor();
       var runsTest = constructor.newInstance();
       if (!runsTest.run(testSuite)) {
-        exit(2);
+        exit(systemExitToggle, 2);
       }
     } catch (ReflectiveOperationException e) {
       e.printStackTrace(System.err);
       System.err.println("Unable to create delegate test runner");
-      exit(2);
+      exit(systemExitToggle, 2);
     }
 
     // Exit manually. If we don't do this then tests which hold resources
     // such as Threads may prevent us from exiting properly.
-    exit(0);
+    exit(systemExitToggle, 0);
+  }
+
+  private static SystemExitToggle getSystemExitToggle() {
+    var version = Runtime.version();
+
+    if (version.feature() < 12) {
+      return new Java11SystemExitToggle();
+    }
+
+    // Load the java 17 toggle by reflection, because it's tied
+    // so closely to the OpenJDK
```

Comment:
```suggestion
    // Load the java 17 toggle by reflection, because it's tied
    // so closely to the OpenJDK.
    // If we import it, it will force `sun.misc.Unsafe` to be class-loaded, which wouldn't work on non-OpenJDKs.
```

Reply:
We also crack open `System` in a way that relies on the OpenJDK implementation of that class. I'll add more text to the comment.